tonic-examples/routeguide-tutorial.md

```diff
@@ -0,0 +1,833 @@
+  rpc RecordRoute(stream Point) returns (RouteSummary) {}
+```
+
+- A *bidirectional streaming RPC* where both sides send a sequence of messages. The two streams
+operate independently, so clients and servers can read and write in whatever
+order they like: for example, the server could wait to receive all the client messages before 
+writing its responses, or it could alternately read a message then write a message, or some other
+combination of reads and writes. The order of messages in each stream is preserved. You specify
+this type of method by placing the `stream` keyword before both the request and the response.
+```proto
+  // Accepts a stream of RouteNotes sent while a route is being traversed,
+  // while receiving other RouteNotes (e.g. from other users).
+  rpc RouteChat(stream RouteNote) returns (stream RouteNote) {}
+```
+
+Our `.proto` file also contains protocol buffer message type definitions for all the request and 
+response types used in our service methods - for example, here's the `Point` message type:
+```proto
+// Points are represented as latitude-longitude pairs in the E7 representation
+// (degrees multiplied by 10**7 and rounded to the nearest integer).
+// Latitudes should be in the range +/- 90 degrees and longitude should be in
+// the range +/- 180 degrees (inclusive).
+message Point {
+  int32 latitude = 1;
+  int32 longitude = 2;
+}
+```
+
+[routeguide-proto]: https://github.com/hyperium/tonic/blob/master/tonic-examples/proto/routeguide/route_guide.proto
+
+## Generating client and server code
+
+Tonic can be configured to generate code as part cargo's normal build process. This is very
+convenient because once we've set everything up, there is no extra step to keep the generated code
+and our `.proto` definitions in sync.
+
+Behind the scenes, Tonic uses [PROST!] to handle protocol buffer serialization and code
+generation.
+
+Edit `Cargo.toml` and add all the dependencies we'll need for this example:
+
+```toml
+[dependencies]
+async-stream = "0.1.2"
+bytes = "0.4"
+futures-preview = { version = "0.3.0-alpha.19", default-features = false, features = ["alloc"]}
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+prost = "0.5"
+rand = "0.7.2"
+tokio = "0.2.0-alpha.6"
+tonic = "0.1.0-alpha.3"
+
+[build-dependencies]
+tonic-build = "0.1.0-alpha.3"
+```
+
+Create a `build.rs` file at the root of your crate:
+
+```rust
+fn main() {
+    tonic_build::compile_protos("proto/route_guide.proto")
+        .unwrap_or_else(|e| panic!("Failed to compile protos {:?}", e));
+}
+```
+
+```shell
+$ cargo build
+```
+
+That's it. The generated code contains:
+
+- Struct definitions for message types `Point`, `Rectangle`, `Feature`, `RouteNote`, `RouteSummary`.
+- A service trait we'll need to implement: `server::RouteGuide`.
+- A client type we'll use to call the server: `client::RouteGuideClient<T>`.
+
+If your are curious as to where the generated files are, keep reading. The mystery will be revealed.
```

Comment:
The sentence "The mystery will be revealed" ends a little abruptly. Is this necessary here? Something like, "I'll be covered soon" or "The mystery will be revealed soon" reads a little better.

Reply:
Done